public/app/features/commandPalette/useMatches.ts

```diff
@@ -0,0 +1,256 @@
+        priority: action.priority + score,
+        action,
+      });
+    }
+
+    ordered = list.sort(order).map((group) => ({
+      name: group.name,
+      actions: map[group.name].sort(order).map((item) => item.action),
+    }));
+
+    /**
+     * Our final result is simply flattening the ordered list into
+     * our familiar (ActionImpl | string)[] shape.
+     */
+    let results: Array<string | ActionImpl> = [];
+    for (let i = 0; i < ordered.length; i++) {
+      let group = ordered[i];
+      if (group.name !== NO_GROUP.name) {
+        results.push(group.name);
+      }
+      for (let i = 0; i < group.actions.length; i++) {
+        results.push(group.actions[i]);
+      }
+    }
+    return results;
+  }, [matches]);
+
+  // ensure that users have an accurate `currentRootActionId`
+  // that syncs with the throttled return value.
+  // eslint-disable-next-line react-hooks/exhaustive-deps
+  const memoRootActionId = React.useMemo(() => rootActionId, [results]);
+
+  return React.useMemo(
+    () => ({
+      results,
+      rootActionId: memoRootActionId,
+    }),
+    [memoRootActionId, results]
+  );
+}
+
+type Match = {
+  action: ActionImpl;
+  /**
+   * Represents the commandScore matchiness value which we use
+   * in addition to the explicitly set `action.priority` to
+   * calculate a more fine tuned fuzzy search.
+   */
+  score: number;
+};
+
+function useInternalMatches(filtered: ActionImpl[], search: string): Match[] {
+  const ufuzzy = useUfuzzy();
+
+  const value = React.useMemo(
+    () => ({
+      filtered,
+      search,
+    }),
+    [filtered, search]
+  );
+
+  const { filtered: throttledFiltered, search: throttledSearch } = useThrottledValue(value);
+
+  return React.useMemo(() => {
+    if (throttledSearch.trim() === '') {
+      return throttledFiltered.map((action) => ({ score: 0, action }));
+    }
+
+    const haystack = throttledFiltered.map((action) =>
+      [action.name, action.keywords, action.subtitle].join(' ').toLowerCase()
+    );
+
+    const results: Match[] = [];
+
+    // If the search term is too long, then just do a simple substring search.
+    // We don't expect users to actually hit this frequently, but want to prevent browser hangs
+    if (throttledSearch.length > FUZZY_SEARCH_LIMIT) {
+      const query = throttledSearch.toLowerCase();
+
+      for (let haystackIndex = 0; haystackIndex < haystack.length; haystackIndex++) {
+        const haystackItem = haystack[haystackIndex];
+
+        // Use the position of the match as a stand-in for score
+        const substringPosition = haystackItem.toLowerCase().indexOf(query);
+
+        if (substringPosition > -1) {
+          const score = haystack.length - substringPosition;
+          const action = throttledFiltered[haystackIndex];
+          results.push({ score, action });
+        }
+      }
+    } else {
+      const allMatchedIndexes = new Set<number>();
+
+      const queryWords = ufuzzy.split(throttledSearch);
```

Comment:
you may want to add a guard here for or 4 or 5 terms before falling back to substring match. permutations are factorial, so it can explode for stuff like `a b c d e f g h i` before our 25 char limit.

Reply:
maybe only do `permutations` if <= 5 words or something?

Reply:
that works too